reefperf/theater.py

```diff
@@ -1,23 +1,36 @@
 from collections import defaultdict
 
-from reefperf.cloud_driver import DRIVER_CLASSES
+from reefperf.driver_manager import DriverManager
+from reefperf.generators.parameters_generator import CreateNodeParametersGenerator
 
 
 class Theater(object):
     def __init__(self, node_types_config, app_deployment_config):
         self._node_types_config = node_types_config
         self._app_deployment_config = app_deployment_config
         self._app_nodes = self._create_app_nodes()
-        self._test_nodes = defaultdict(list)
+
+    def app_nodes_by_type(self, node_type):
+        return self._app_nodes[node_type]
 
     def _create_app_nodes(self):
         nodes = defaultdict(list)
-        for node_deploy_config in self._app_deployment_config["app-deployment"]["nodes"]:
-            node_type = node_deploy_config["type"]
-            node_config = self._node_types_config["app-node-types"][node_type]
-            node_config["name"] = node_deploy_config["name"]
-            driver_class = DRIVER_CLASSES[node_config["driver-class"]]
-            driver = driver_class()
-            node = driver.create_node(node_config)
-            nodes[node_deploy_config["type"]].append(node)
+        for node_type, node_deploy_config in self._app_deployment_config["app_deployment"]["nodes"].items():
+            node_num = node_deploy_config.get("count", 1)
+            for _ in range(node_num):
+                node = self._create_node(node_type)
+                nodes[node_type].append(node)
         return nodes
+
+    def _create_node(self, node_type):
+        node_deploy_config = self._app_deployment_config["app_deployment"]["nodes"][node_type]
+        driver_manager = DriverManager.get_instance()
+        driver_class = self._node_types_config["app_node_types"][node_type]["driver_class"]
+        driver_parameters = self._node_types_config["app_node_types"][node_type]["create_driver_parameters"]
+        driver = driver_manager.get_driver(driver_class, **driver_parameters)
```

Comment:
we construct a new CouldDriver every time we want to create a node? That will securely authorize us with the cloud API etc, can take seconds. If we want to create a hundred nodes, it's going to cost us a lot of money. Maybe we can re-use the driver between the node creations (and even between different spectacles)?

Reply:
`DriverManager`'s `get_driver` method has @lru_cache decorator. So, I've already implemented this strategy